### PR TITLE
World generator

### PR DESCRIPTION
**Pull request by nicpalard**
Adding Randomizer Object
Adding WorldGenerator Object
Adding test for the randomizer.

### CODE REVIEW

File: tests/Randomizer.cpp [hunk 1]
```diff
@@ -18,227 +18,363 @@
...
 
-	void addRobots(World* world)
+	return Point(posx, posy);
+}
+
+Color Randomizer::randColor()
+{
+	float r = randColorFloat();
+	float g = randColorFloat();
+	float b = randColorFloat();
```
Comment (Krazyxx) [inline tests/Randomizer.cpp]: Where is the alpha channel value?

File: tests/randomWorld.cpp [hunk 2]
```diff
@@ -32,80 +32,207 @@ using namespace Enki;
...
-			REQUIRE( (thymio->pos.x <= w->w && thymio->pos.y <= w->h) );
-		// TODO: else with CIRCULAR WALLS (Require to compute possible positions
-		// from circle radius)
+	SECTION ( "A random color" ) {
+		for (int i = 0; i < ITERATION_NUMBER; i++)
+		{
+			Color c = globalRandomizer->randColor();
+			REQUIRE ( (c.r() >= 0 && c.r() <= 1) );
+			REQUIRE ( (c.g() >= 0 && c.g() <= 1) );
+			REQUIRE ( (c.b() >= 0 && c.b() <= 1) );
```
Comment (Krazyxx) [inline tests/randomWorld.cpp]: Need a REQUIRE for the alpha channel value.

File: tests/randomWorld.cpp [hunk 3]
```diff
@@ -32,80 +32,207 @@ using namespace Enki;
...
+			REQUIRE ( (w->objects.size() == 0) );
+
+			delete w;
 		}
-		else
+	}
+
+	SECTION ( "A random Robot" ) {
+		globalRandomizer->randWorld();
+		World* w =globalRandomizer->getWorld();
```
Comment (Krazyxx) [inline tests/randomWorld.cpp]: @nicpalard ping

File: tests/Randomizer.cpp [hunk 4]
```diff
@@ -18,227 +18,363 @@
...
-#include <chrono>
 
-/*!	\file Randomizer.cpp
-	\brief Module 'random' which can be used for generating objects and worlds.
-*/
+using namespace Enki;
+
+Randomizer::Randomizer(World* world, long long int seed)
+{
+
```
Comment (Krazyxx) [inline tests/Randomizer.cpp]: @nicpalard ping

File: tests/Randomizer.cpp [hunk 5]
```diff
@@ -18,227 +18,363 @@
...
 	}
 
-	EPuck* randomEPuck(Enki::World* w)
+	return hull;
+}
+
+PhysicalObject::Part Randomizer::randComplexPart()
+{
+	Polygone p = randConvexPolygone(randInt(3, 30));
+	bool textured = randBool();
```
Comment (Krazyxx) [inline tests/Randomizer.cpp]: Prefer "isTextured" ! :+1: :dango:

File: tests/Randomizer.cpp [hunk 6]
```diff
@@ -18,227 +18,363 @@
...
+{
+	float size1 = randFloat(1.0, 30.0), size2 = randFloat(1.0, 30.0), height = randFloat(1.0, 30.0);
+	return PhysicalObject::Part(size1, size2, height);
+}
+
+Polygone Randomizer::randConvexPolygone(int polygoneSize)
+{
+	polygoneSize = polygoneSize < 3 ?
+		3 :
+		polygoneSize;
```
Comment (Krazyxx) [inline tests/Randomizer.cpp]: @nicpalard ping

File: tests/Randomizer.cpp [hunk 7]
```diff
@@ -18,227 +18,363 @@
...
+		posx = randFloat(0, this->world->w);
+		posy = randFloat(0, this->world->h);
 	}
-
-	void addObjects(World* world)
+	else if (this->world->wallsType == World::WALLS_CIRCULAR)
 	{
-		int objectsNumber = randomNumber(1, MAX_DYNAMIC_OBJECTS);
-		for (int i = 0; i < objectsNumber; i++)
+		bool in = 0;
```
Comment (Krazyxx) [inline tests/Randomizer.cpp]: @nicpalard ping

File: tests/WorldGenerator.cpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "WorldGenerator.h"
+
+namespace Enki
+{
+
```
Comment (Krazyxx) [inline tests/WorldGenerator.cpp]: @nicpalard ping

File: tests/Randomizer.h [hunk 9]
```diff
@@ -16,140 +16,234 @@
...
+
+		/*!
+			\brief Constructor of a Randomizer.
+			A randomizer aims to generate random Enki Worlds and Enki
+			Objects.
+			This constructor will generate a random world because no
+			base world is specified.
+		*/
+		Randomizer(long long int seed = -1);
+
```
Comment (Krazyxx) [inline tests/Randomizer.h]: @nicpalard ping

File: tests/Randomizer.cpp (hunk 6 above)
Comment (sebpouteau) [inline tests/Randomizer.cpp]: why ping? this syntax is correct

File: tests/randomWorld.cpp [hunk 10]
```diff
@@ -32,80 +32,207 @@ using namespace Enki;
...
+	// If you want to see if it correctly rands random numbers
+	// You should check the first SECTION.
+	SECTION ( "A random position in the world" ) {
+		World* w = globalRandomizer->getWorld();
+		for (int i = 0; i < ITERATION_NUMBER; i++)
 		{
-			REQUIRE( (w->r <= MAX_RADIUS && w->r >= MIN_RADIUS) );
-			REQUIRE( w->objects.size() <= (MAX_DYNAMIC_OBJECTS + MAX_ROBOTS) );
-			if (DEBUG)
+			Point p =globalRandomizer->randPoint();
```
Comment (Krazyxx) [inline tests/randomWorld.cpp]: @nicpalard ping

File: tests/randomWorld.cpp [hunk 11]
```diff
@@ -32,80 +32,207 @@ using namespace Enki;
...
+			REQUIRE ( (p.getHeight() >= 1 && p.getHeight() <= 30) );
+			REQUIRE ( (p.getArea() <= (30 * 30)) );
+		}
+	}
+
+	SECTION ( "A random texture" ) {
+		for (int i = 0; i < ITERATION_NUMBER; i++)
+		{
+			Texture t = globalRandomizer->randTexture();
+			REQUIRE ( (t.size() >= 1 && t.size() <= 5) );
```
Comment (Krazyxx) [inline tests/randomWorld.cpp]: @nicpalard ping : `REQUIRE ( (1 <= t.size() && t.size() <= 5) );` → better visibility.

File: tests/randomWorld.cpp [hunk 12]
```diff
@@ -32,80 +32,207 @@ using namespace Enki;
...
+					(w->w >= MIN_WIDTH && w->w <= MAX_WIDTH) ) );
 			}
+			else
+			{
+				REQUIRE ( ((w->r >= MIN_RADIUS && w->r <= MAX_RADIUS) &&
+				w->h == 0 &&
+				w->w == 0) );
+			}
+
+			REQUIRE ( (w->objects.size() == 0) );
```
Comment (Krazyxx) [inline tests/randomWorld.cpp]: @nicpalard ping

File: tests/randomWorld.cpp [hunk 13]
```diff
@@ -32,80 +32,207 @@ using namespace Enki;
...
 		}
-		// Get this out of if to prevent duplication
-		REQUIRE( w->objects.size() <= (MAX_DYNAMIC_OBJECTS + MAX_ROBOTS) );
-		if (DEBUG)
-			cout << "- number of objects : " << w->objects.size() << endl;
-		for (PhysicalObject* p : w->objects)
+	}
+
+	SECTION ( "A random physical object" ) {
+
```
Comment (Krazyxx) [inline tests/randomWorld.cpp]: @nicpalard ping

File: tests/serialization.cpp [hunk 14]
```diff
@@ -22,9 +22,9 @@
 #include <enki/PhysicalEngine.h>
 #include <enki/robots/thymio2/Thymio2.h>
 #include <enki/Serialize.h>
-#include "Randomizer.h"
+#include "WorldGenerator.h"
 
-const double EPSILON = 0.0001;
+const double EPSILON = 0.001;
```
Comment (Krazyxx) [inline tests/serialization.cpp]: @nicpalard ping :taco:

File: tests/serialization.cpp [hunk 15]
```diff
@@ -75,6 +75,8 @@ static bool equalsThymio(Thymio2* t1, Thymio2* t2)
 	if (!equalsPoint(t1->pos, t2->pos))
 	{
 		cerr << "[T] Not same position" << endl;
+		cerr << "  [POS1]: " << t1->pos.x << "x" << t1->pos.y << endl;
+		cerr << "  [POS2]: " << t2->pos.x << "x" << t2->pos.y << endl;
```
Comment (Krazyxx) [inline tests/serialization.cpp]: @nicpalard ping

File: tests/serialization.cpp [hunk 16]
```diff
@@ -198,9 +201,20 @@ TEST_CASE( "Serialization", "[Serialization Reproducibility]" ) {
 	SECTION( "[S] World with one Thymio") {
 		for (int i = 0; i < ITERATION_NUMBER; i++)
 		{
-			World* w = randomWorld();
-			Thymio2* t = randomThymio(w);
-			w->addObject(t);
+			std::cout << "World with one Thymio" << std::endl;
+			// Since it's looping, we need to reset the world
+			gen.resetWorld();
+			// Adding 1 thymio2 to the current generated world
```
Comment (Krazyxx) [inline tests/serialization.cpp]: @nicpalard ping : useless comment :walking_man:

File: tests/serialization.cpp [hunk 17]
```diff
@@ -198,9 +201,20 @@ TEST_CASE( "Serialization", "[Serialization Reproducibility]" ) {
...
+			gen.add(Randomizer::THYMIO2_, 1);
+			World* w = gen.getWorld();
+
+			for(auto& obj : w->objects)
+			{
+				std::cerr << "OBJ: " << typeid(*obj).name() << std::endl;
+			}
+
```
Comment (Krazyxx) [inline tests/serialization.cpp]: @nicpalard ping : extra space :call_me_hand:

File: tests/Randomizer.cpp (hunk 6 above)
Comment (nicpalard) [inline tests/Randomizer.cpp]: He doesn't seem to like the multiple line ternary expression

File: tests/Randomizer.cpp [hunk 18]
```diff
@@ -113,16 +109,17 @@ PhysicalObject* Randomizer::randPhysicalObject(const int &hullSize)
...
 	if (cylindric)
 	{
 		object->setCylindric(randFloat(1.0, 5.0), randFloat(1.0, 5.0), randFloat(1.0, 5.0));
 	}
 	else
 	{
 		// This generates a bug that moves the object out of the world
-		// when a complex part is computed with a big height.
+		// when a complex part is computed with a big height and the object
+		// is near the world's borders.
```
Comment (nicpalard) [inline tests/Randomizer.cpp]: @nicpalard fix this.

File: enki/worldgenerator/Randomizer.cpp [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+	this->world = world;
+	std::cerr << "seed: " << this->seed << std::endl;
+}
+
+Randomizer::Randomizer(const long long int &seed)
+{
+	/*
+	Since c++11 you can call something like "super" constructor.
+	Something like : Randomizer(randomizeWorld(), seed); would
+	be more convinient but this doesn't work as the seed needs to be set
```
Comment (mthl) [inline enki/worldgenerator/Randomizer.cpp]: convinient => convenient

File: enki/worldgenerator/Randomizer.h [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+	private:
+		World* world;
+		long long int seed;
+
+		/* We use Mersenne Twister 19937 engine because
+		it's the most popular engine for
+		PRNG (Pseudo random number generation)
+		Mersenne Twister engine procude better output
+		than the default engine but it is way slower
+		Also the default_randome_engine behavior isn't fixed a it directly depends on the implementation of the libc++ which can change depending on the OS.
```
Comment (mthl) [inline enki/worldgenerator/Randomizer.h]: libc => libstdc++

File: enki/worldgenerator/WorldGenerator.cpp [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+		{
+			PhysicalObject* o;
+			/*
+			type can have up to 7 values :
+				0 -> 4 : Only robots (of a specified type) (enum)
+				5 : Physical Object
+				6 : Any kind of robots
+				7 : Anything (PO / R)
+			*/
+			switch (type) {
```
Comment (mthl) [inline enki/worldgenerator/WorldGenerator.cpp]: newline before `{`

**Pull request merged**